File: fsharp/bank-account/BankAccount.fs
module BankAccount

type BankAccount() =
    let lockObj = obj()
    member val Closed = true
        with get, set
    member val Balance = 0.0m
        with get, set
    member this.Update(change: decimal) =
        lock lockObj (fun () -> 
            this.Balance <- this.Balance + change)

let mkBankAccount(): BankAccount = BankAccount() 

let openAccount (account: BankAccount): BankAccount = 
    account.Closed <- false
    account

let closeAccount (account: BankAccount): BankAccount = 
    account.Closed <- true 
    account

let getBalance (account: BankAccount): decimal option = 
    match account.Closed with
    | true -> None
    | false -> account.Balance |> Some

let updateBalance (change: decimal) (account: BankAccount) = 
    account.Update(change)
    account

File: csharp/bank-account/BankAccount.cs
using System;

public class BankAccount
{
    private bool _isClosed = false;
    private decimal _balance = 0m;
    public void Open()
    {
        return;
    }

    public void Close()
    {
        _isClosed = true;
    }

    public decimal Balance
    {
        get
        {
            if (_isClosed) throw new InvalidOperationException();
            return _balance;
        }
        private set
        {
            _balance = value;
        }
    }

    public void UpdateBalance(decimal change)
    {
        lock(this)
        {
            Balance += change;
        }
        return;
    }
}
